frostsnap_core/src/message.rs

```diff
@@ -263,3 +278,39 @@ pub struct SignItem {
     pub tap_tweak: bool,
     pub bip32_path: Vec<u32>,
 }
+
+impl SignItem {
+    pub fn derive_key<K: TweakableKey>(&self, root_key: &K) -> K::XOnly {
+        let mut xpub = crate::xpub::Xpub::new(root_key.clone());
+        xpub.derive_bip32(&self.bip32_path);
+
+        if self.tap_tweak {
+            let tweak = bitcoin::util::taproot::TapTweakHash::from_key_and_tweak(
+                xpub.key().to_libsecp_xonly(),
+                None,
+            )
+            .to_scalar();
+            xpub.key().clone().into_xonly_with_tweak(
+                Scalar::<Public, _>::from_bytes_mod_order(tweak.to_be_bytes())
+                    .non_zero()
+                    .expect("computationally unreachable"),
+            )
+        } else {
+            xpub.key().clone().into_xonly()
+        }
+    }
```

Comment:
I realize it was like this before, but it's hard to follow the reasoning for these steps.

We take our root Frostsnap key and derive an extended public key by adding tweaks according to the BIP32 path.

Then, if the key is tweaked, we get the TapTweakHash from this extended public key. A "cumulative tweak" from the derived key?

Then using this tweak, we add take our derived extended public key again and finally convert it to an xonly key under this tweak.

The code makes it seem like we are adding the tweak in multiple times.

Should self.tap_tweak always be true if self.bip32_path is not empty?

Reply:
> I realize it was like this before, but it's hard to follow the reasoning for these steps.
> 
> We take our root Frostsnap key and derive an extended public key by adding tweaks according to the BIP32 path.
> 
> Then, if the key is tweaked, we get the TapTweakHash from this extended public key. A "cumulative tweak" from the derived key?

There is no cumulative tweak. The TapTweakHash is not from the extended public key. Just from the xonly key. I've pushed a commit to make this clearer as they were being conflated before: 45e2dc816abd5482a63d65fd21face69240ac53d

> 
> Then using this tweak, we add take our derived extended public key again and finally convert it to an xonly key under this tweak.
> 
> The code makes it seem like we are adding the tweak in multiple times.

we are.

> 
> Should `self.tap_tweak` always be true if `self.bip32_path` is not empty?

They're orthogonal. You can do a BIP32 derivation for a nostr key for example.

Reply:
Ah got it, thanks!